File: backend/models/Task.js
// ملف: models/Task.js
// الغرض: تعريف نموذج المهمة وإدارة خصائصها وعلاقاتها مع المستخدم والفئات
// الحقول الرئيسية: العنوان، الوصف، الأولوية، الحالة، التواريخ، مكافأة XP، التكرار، العلامات
// العلاقات:
// - Task ينتمي إلى User (مطلوب)
// - Task ينتمي إلى Category (اختياري)
// Hooks:
// - beforeUpdate: تعيين completedAt تلقائياً عند تغيير الحالة إلى completed

const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Task = sequelize.define('Task', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  userId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    field: 'user_id'
  },
  categoryId: {
    type: DataTypes.INTEGER,
    field: 'category_id'
  },
  title: {
    type: DataTypes.STRING(255),
    allowNull: false,
    validate: {
      len: [1, 255]
    }
  },
  description: {
    type: DataTypes.TEXT
  },
  priority: {
    type: DataTypes.ENUM('low', 'medium', 'high', 'urgent'),
    defaultValue: 'medium'
  }, // ENUM: ترتيب الأولوية (low/medium/high/urgent)
  status: {
    type: DataTypes.ENUM('pending', 'in_progress', 'completed', 'cancelled'),
    defaultValue: 'pending'
  }, // ENUM: حالة المهمة (pending/in_progress/completed/cancelled)
  dueDate: {
    type: DataTypes.DATE,
    field: 'due_date'
  },
  reminderDate: {
    type: DataTypes.DATE,
    field: 'reminder_date'
  },
  estimatedDuration: {
    type: DataTypes.INTEGER, // in minutes
    field: 'estimated_duration'
  }, // مدة التنفيذ المقدرة بالدقائق
  actualDuration: {
    type: DataTypes.INTEGER, // in minutes
    field: 'actual_duration'
  }, // مدة التنفيذ الفعلية بالدقائق
  xpReward: {
    type: DataTypes.INTEGER,
    defaultValue: 10,
    field: 'xp_reward'
  },
  isRecurring: {
    type: DataTypes.BOOLEAN,
    defaultValue: false,
    field: 'is_recurring'
  }, // تكرار المهمة
  recurringPattern: {
    type: DataTypes.STRING(50),
    field: 'recurring_pattern'
  }, // نمط التكرار (مثال: daily/weekly)
  tags: {
    type: DataTypes.JSON,
    defaultValue: []
  },
  attachments: {
    type: DataTypes.JSON,
    defaultValue: []
  },
  completedAt: {
    type: DataTypes.DATE,
    field: 'completed_at'
  }
}, {
  tableName: 'tasks',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at',
  hooks: {
    // عند تحديث الحالة إلى "completed" لأول مرة، يتم تعيين تاريخ الإكمال
    beforeUpdate: (task) => {
      if (task.changed('status') && task.status === 'completed' && !task.completedAt) {
        task.completedAt = new Date();
      }
    }
  }
});

// Instance Methods
// markCompleted(): تحديد المهمة كمكتملة وتعيين وقت الإكمال
Task.prototype.markCompleted = async function() {
  this.status = 'completed';
  this.completedAt = new Date();
  await this.save();
  return this;
};

// isOverdue(): التحقق من تأخر المهمة (إن كان dueDate في الماضي والحالة ليست مكتملة)
Task.prototype.isOverdue = function() {
  if (!this.dueDate || this.status === 'completed') return false;
  return new Date() > new Date(this.dueDate);
};

// getDaysUntilDue(): حساب عدد الأيام المتبقية حتى تاريخ الاستحقاق (أو null إن لم يكن محدداً)
Task.prototype.getDaysUntilDue = function() {
  if (!this.dueDate) return null;
  const today = new Date();
  const due = new Date(this.dueDate);
  const diffTime = due - today;
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
};

// getPriorityScore(): تحويل الأولوية إلى رقم للمقارنة (urgent=4, high=3, medium=2, low=1)
Task.prototype.getPriorityScore = function() {
  const priorityScores = {
    low: 1,
    medium: 2,
    high: 3,
    urgent: 4
  };
  return priorityScores[this.priority] || 2;
};

// calculateXpReward(): حساب مكافأة XP بناءً على الأولوية والمدة المقدرة والتأخير
// الصيغة: baseXp + priorityBonus + durationBonus - overduePenalty (الحد الأدنى 5 XP)
Task.prototype.calculateXpReward = function() {
  let baseXp = 10;
  
  // Priority bonus
  const priorityBonus = {
    low: 0,
    medium: 5,
    high: 10,
    urgent: 20
  };
  
  // Duration bonus (for longer tasks)
  const durationBonus = this.estimatedDuration ? Math.floor(this.estimatedDuration / 30) * 5 : 0;
  
  // Overdue penalty
  const overduePenalty = this.isOverdue() ? -5 : 0;
  
  this.xpReward = Math.max(baseXp + priorityBonus[this.priority] + durationBonus + overduePenalty, 5);
  return this.xpReward;
};

module.exports = Task;

File: frontend/lib/core/models/task.dart
import 'package:hive/hive.dart';

part 'task.g.dart';

/// نموذج المهمة (Task Model)
/// 
/// يمثل مهمة واحدة في التطبيق مع جميع خصائصها:
/// - المعلومات الأساسية (العنوان، الوصف)
/// - الأولوية والحالة
/// - التواريخ (الإنشاء، الاستحقاق، الإكمال)
/// - نظام التحفيز (XP Reward)
/// - التكرار والعلامات
/// 
/// يستخدم Hive للتخزين المحلي مع TypeAdapter
@HiveType(typeId: 0)
class Task extends HiveObject {
  /// معرف المهمة الفريد
  @HiveField(0)
  String id;

  /// عنوان المهمة
  @HiveField(1)
  String title;

  /// وصف المهمة (اختياري)
  @HiveField(2)
  String? description;

  /// أولوية المهمة: low, medium, high, urgent
  @HiveField(3)
  String priority; // low, medium, high, urgent

  /// حالة المهمة: pending, in_progress, completed, cancelled
  @HiveField(4)
  String status; // pending, in_progress, completed, cancelled

  /// تاريخ الاستحقاق (اختياري)
  @HiveField(5)
  DateTime? dueDate;

  /// تاريخ ووقت التذكير (اختياري)
  @HiveField(6)
  DateTime? reminderDate;

  /// المدة المقدرة بالدقائق (اختياري)
  @HiveField(7)
  int? estimatedDuration; // in minutes

  /// المدة الفعلية بالدقائق (اختياري)
  @HiveField(8)
  int? actualDuration; // in minutes

  /// نقاط الخبرة (XP) المكتسبة عند الإكمال
  @HiveField(9)
  int xpReward;

  /// هل المهمة متكررة؟
  @HiveField(10)
  bool isRecurring;

  /// نمط التكرار (مثال: daily/weekly) إن وجِد
  @HiveField(11)
  String? recurringPattern;

  /// العلامات المرتبطة بالمهمة
  @HiveField(12)
  List<String> tags;

  /// معرف الفئة المرتبطة (اختياري)
  @HiveField(13)
  String? categoryId;

  /// وقت الإكمال (يُعين عند اكتمال المهمة)
  @HiveField(14)
  DateTime? completedAt;

  /// وقت الإنشاء
  @HiveField(15)
  DateTime createdAt;

  /// آخر وقت تحديث
  @HiveField(16)
  DateTime updatedAt;

  Task({
    required this.id,
    required this.title,
    this.description,
    this.priority = 'medium',
    this.status = 'pending',
    this.dueDate,
    this.reminderDate,
    this.estimatedDuration,
    this.actualDuration,
    this.xpReward = 10,
    this.isRecurring = false,
    this.recurringPattern,
    this.tags = const [],
    this.categoryId,
    this.completedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  /// تحويل المهمة إلى JSON للإرسال للـ API
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'priority': priority,
      'status': status,
      'dueDate': dueDate?.toIso8601String(),
      'reminderDate': reminderDate?.toIso8601String(),
      'estimatedDuration': estimatedDuration,
      'actualDuration': actualDuration,
      'xpReward': xpReward,
      'isRecurring': isRecurring,
      'recurringPattern': recurringPattern,
      'tags': tags,
      'categoryId': categoryId,
      'completedAt': completedAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  /// إنشاء مهمة من JSON المستلم من الـ API
  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      priority: json['priority'] ?? 'medium',
      status: json['status'] ?? 'pending',
      dueDate: json['dueDate'] != null ? DateTime.parse(json['dueDate']) : null,
      reminderDate: json['reminderDate'] != null ? DateTime.parse(json['reminderDate']) : null,
      estimatedDuration: json['estimatedDuration'],
      actualDuration: json['actualDuration'],
      xpReward: json['xpReward'] ?? 10,
      isRecurring: json['isRecurring'] ?? false,
      recurringPattern: json['recurringPattern'],
      tags: List<String>.from(json['tags'] ?? []),
      categoryId: json['categoryId'],
      completedAt: json['completedAt'] != null ? DateTime.parse(json['completedAt']) : null,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : DateTime.now(),
    );
  }

  /// إنشاء نسخة من المهمة مع تعديل بعض الحقول
  Task copyWith({
    String? id,
    String? title,
    String? description,
    String? priority,
    String? status,
    DateTime? dueDate,
    DateTime? reminderDate,
    int? estimatedDuration,
    int? actualDuration,
    int? xpReward,
    bool? isRecurring,
    String? recurringPattern,
    List<String>? tags,
    String? categoryId,
    DateTime? completedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      priority: priority ?? this.priority,
      status: status ?? this.status,
      dueDate: dueDate ?? this.dueDate,
      reminderDate: reminderDate ?? this.reminderDate,
      estimatedDuration: estimatedDuration ?? this.estimatedDuration,
      actualDuration: actualDuration ?? this.actualDuration,
      xpReward: xpReward ?? this.xpReward,
      isRecurring: isRecurring ?? this.isRecurring,
      recurringPattern: recurringPattern ?? this.recurringPattern,
      tags: tags ?? this.tags,
      categoryId: categoryId ?? this.categoryId,
      completedAt: completedAt ?? this.completedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? DateTime.now(),
    );
  }

  // Helper methods
  /// هل المهمة مكتملة؟ (status == 'completed')
  bool get isCompleted => status == 'completed';
  /// هل المهمة متأخرة؟ (dueDate في الماضي وغير مكتملة)
  bool get isOverdue => dueDate != null && DateTime.now().isAfter(dueDate!) && !isCompleted;
  
  /// عدد الأيام المتبقية حتى موعد الاستحقاق
  int get daysUntilDue {
    if (dueDate == null) return 0;
    return dueDate!.difference(DateTime.now()).inDays;
  }

  /// تحويل الأولوية إلى قيمة رقمية للترتيب (urgent=4, high=3, medium=2, low=1)
  int get priorityScore {
    switch (priority) {
      case 'urgent':
        return 4;
      case 'high':
        return 3;
      case 'medium':
        return 2;
      case 'low':
        return 1;
      default:
        return 2;
    }
  }
}
